### ISSUE TITLE
Improve level constraint solving for groups of inequality constraints

### ISSUE THREAD

**Issue by nad** [agda/agda#2780]
Consider the following code:
```agda
open import Agda.Primitive

postulate
  ℓ : Level
  F : Level → Set

module Two where

  mutual

    α : Level
    α = _

    G : F (α ⊔ ℓ) → F ℓ
    G A = A

    H : F (α ⊔ ℓ) → F α
    H A = A

module Three where

  mutual

    α : Level
    α = _

    β : Level
    β = _

    G : F (α ⊔ β ⊔ ℓ) → F ℓ
    G A = A

    H : F (α ⊔ β ⊔ ℓ) → F α
    H A = A

    I : F (α ⊔ β ⊔ ℓ) → F β
    I A = A
```
Agda solves the meta-variable in `Two`, but not the metas in `Three`. Error message:
```
Failed to solve the following constraints:
  _16 ⊔ (_17 ⊔ ℓ) = _17
  _16 ⊔ (_17 ⊔ ℓ) = _16
  _16 ⊔ (_17 ⊔ ℓ) = ℓ
```
These equations have a unique solution, `_16 = _17 = ℓ`. In `Two` I guess that Agda solves (more or less) the following equations successfully:
```
  α ⊔ ℓ = ℓ
  α ⊔ ℓ = α
```
Why can Agda solve the meta in `Two`, but not the metas in `Three`?

**Comment by UlfNorell** [comment]
There's special code for the `Two` case in [Agda.TypeChecking.LevelConstraints](https://github.com/agda/agda/blob/0f817be40775d56a303565689ee4a6e3787e1c71/src/full/Agda/TypeChecking/LevelConstraints.hs#L14-L21).

**Comment by nad** [comment]
What is the rationale for having a special case (AKA a hack)?

Speculation: Constraints of the form xᵢ = x₁ ⊔ … ⊔ xᵢ ⊔ … ⊔ xₙ mean exactly the same thing as xᵢ ≥ x₁, …, xᵢ ≥ xₙ (excluding xᵢ ≥ xᵢ). One could transform constraints in this way, and if a pair of constraints xᵢ ≥ xⱼ and xⱼ ≥ xᵢ is found, then these can be turned into xᵢ = xⱼ. This would handle the situation in `Three`.

**Comment by UlfNorell** [comment]
> if a pair of constraints xᵢ ≥ xⱼ and xⱼ ≥ xᵢ is found, then these can be turned into xᵢ = xⱼ

This is exactly what `simplifyLevelConstraint` is doing (I think the technical term is *antisymmetry*, not *hack*), so the only thing you would have to do is to split up the `xᵢ = x₁ ⊔ … ⊔ xᵢ ⊔ … ⊔ xₙ` constraints.

**Comment by andreasabel** [comment]
Seems to be related to #2246.

**Comment by andreasabel** [comment]
Here is another example where transitivity of level constraints is needed:
```agda
mutual
  l = _

  data D {a} (A : Set a) : Set l where
    c : A → D A

  data E (A : Set l) : Set1 where
    c : A → E A
```
Agda rejects this because it cannot solve the constraints, but it should discover that they are definitely unsatisfiable.
```
Failed to solve the following constraints:
  _4 =< .Agda.Primitive.lsuc .Agda.Primitive.lzero
  a =< _4
```
Interestingly, if we replace `Set1` by `Set`, Agda can see that they are unsolvable, since then `_4` can be solved and assigned to zero.

Partial order constraints as we have in levels need to be attacked in conjunction, just as for size constraints, to make use of transitivity.  Currently, Agda tries to attack them locally, one by one, just as for Miller unification constraints, with a few hacks to make some common use cases go through nevertheless.

> Constraints of the form xᵢ = x₁ ⊔ … ⊔ xᵢ ⊔ … ⊔ xₙ mean exactly the same thing as xᵢ ≥ x₁, …, xᵢ ≥ xₙ (excluding xᵢ ≥ xᵢ). 

No, what you say expresses just "upper bound", not "least upper bound".

**Comment by nad** [comment]
> > Constraints of the form xᵢ = x₁ ⊔ … ⊔ xᵢ ⊔ … ⊔ xₙ mean exactly the same thing as xᵢ ≥ x₁, …, xᵢ ≥ xₙ (excluding xᵢ ≥ xᵢ). 
>
> No, what you say expresses just "upper bound", not "least upper bound".

Note that the term xᵢ  occurs on both sides of the equation.

**Comment by UlfNorell** [comment]
This was fixed by c0d124275.

**Comment by andreasabel** [comment]
I filed my example https://github.com/agda/agda/issues/2780#issuecomment-334754540 as #3098.
